project-koku/koku
# [Cost -1861] - adding a cost_type parameter to AWS report api

Adding a cost type parameter to AWS report api to add in the capability of AWS amortized saving plan costs in the provider map. 

#### Test Instructions:
1. pull branch and checkout branch
2. start up docker and load test data
3. hit the following endpoint
```
http://127.0.0.1:8000/api/cost-management/v1/reports/aws/costs/?filter%5Btime_scope_value%5D=-1&check_tags=True

results should include the following lines:
{
    "meta": {
        "count": 10,
        "cost_type": "unblended_cost",
        "compute_count": false,
        "check_tags": false,
        "filter": {
            "time_scope_value": "-10",
            "time_scope_units": "day",
            "resolution": "daily"
        },
```
which shows cost_type set to its default value of "unblended_cost"

4. hit the following endpoint
```
http://127.0.0.1:8000/api/cost-management/v1/reports/aws/costs/?filter%5Btime_scope_value%5D=-1&check_tags=True&cost_type=savingsplan_effective_cost

results should include the following lines:
{
    "meta": {
        "count": 1,
        "cost_type": "savingsplan_effective_cost",
        "compute_count": false,
        "check_tags": true,
        "filter": {
            "time_scope_value": "-1"
        },
```
which shows cost_type set to a filtered value of savingsplan_effective_cost


#### Front-end testing utilizing settings page
1. pull branch and checkout branch
2. start up docker and load test data
3. Load the settings page and leave the cost_type as unblended
4. hit the following endpoint
```
http://127.0.0.1:8000/api/cost-management/v1/reports/aws/costs/?filter%5Btime_scope_value%5D=-1&check_tags=True

results should include the following lines:
{
    "meta": {
        "count": 10,
        "cost_type": "unblended_cost",
        "compute_count": false,
        "check_tags": false,
        "filter": {
            "time_scope_value": "-10",
            "time_scope_units": "day",
            "resolution": "daily"
        },
```
5. Now in the settings page select a new setting for the users cost_type 
6. hit the endpoint
```
http://127.0.0.1:8000/api/cost-management/v1/reports/aws/costs/?filter%5Btime_scope_value%5D=-1&check_tags=True

results should include the following lines:
{
    "meta": {
        "count": 10,
        "cost_type": [cost_type selected in the user settings page],
        "compute_count": false,
        "check_tags": false,
        "filter": {
            "time_scope_value": "-10",
            "time_scope_units": "day",
            "resolution": "daily"
        },
```
7. Also test filtering above adding &cost_type=[savingsplan_effective_cost, unblended_cost, or blended_cost] to ensure the filtering still works with the user_settings table populated

#### Smoke Tests 
Dan and Eva signed off on the smoke tests for this branch. We had the following failures but they were unrelated to our changes: 
```
========================================================================================================== short test summary info ===========================================================================================================
FAILED iqe_cost_management/tests/rest_api/v1/test_resource_type_endpoints.py::test_api_gcp_account_search - assert [] != []
FAILED iqe_cost_management/tests/rest_api/v1/test_resource_type_endpoints.py::test_api_gcp_product_search - assert [] != []
FAILED iqe_cost_management/tests/rest_api/v1/test_resource_type_endpoints.py::test_api_gcp_regions_search - assert [] != []
FAILED iqe_cost_management/tests/rest_api/v1/test_resource_type_endpoints.py::test_api_gcp_services_search - assert [] != []
FAILED iqe_cost_management/tests/rest_api/v1/test_resource_type_endpoints.py::test_api_gcp_accounts_compare_data - AssertionError: assert 'todays_raw_calc_id' in []
FAILED iqe_cost_management/tests/rest_api/v1/test_resource_type_endpoints.py::test_api_gcp_projects_compare_data - assert [] != []
FAILED iqe_cost_management/tests/rest_api/v1/test_resource_type_endpoints.py::test_api_gcp_regions_compare_data - assert [] != []
FAILED iqe_cost_management/tests/rest_api/v1/test_resource_type_endpoints.py::test_api_gcp_services_compare_data - assert [] != []
FAILED iqe_cost_management/tests/rest_api/v1/test_source.py::test_api_gcp_source_raw_calc - AssertionError: Raw calculated value does not match report value
FAILED iqe_cost_management/tests/rest_api/v1/test_source.py::test_api_source_filter_by_type - AssertionError: assert 'AWS-local' == 'AWS'
FAILED iqe_cost_management/tests/rest_api/v1/test_user.py::test_api_settings_azure_tag_enablement - dynaconf.vendor.box.exceptions.BoxKeyError: "'DynaBox' object has no attribute 'PRIMARY_USER'"
FAILED iqe_cost_management/tests/rest_api/v1/test_user.py::test_api_settings_gcp_tag_enablement - dynaconf.vendor.box.exceptions.BoxKeyError: "'DynaBox' object has no attribute 'PRIMARY_USER'"
FAILED iqe_cost_management/tests/rest_api/v1/test_user.py::test_api_settings_aws_tag_enablement - dynaconf.vendor.box.exceptions.BoxKeyError: "'DynaBox' object has no attribute 'PRIMARY_USER'"
=========================================================================== 13 failed, 5231 passed, 15 skipped, 876 deselected, 90 warnings in 2598.85s (0:43:18) ============================================================================
```
